### PR TITLE
refactor secure notebook server script

### PR DESCRIPTION
**Pull request by Carreau**
break-up into  pieces and write docstring.

### CODE REVIEW

**Comment by minrk** [comment]
I like it. cc @fperez who requested the changes.

**Comment by Carreau** [comment]
Also https://github.com/ipython/traitlets/pull/137 might simplify this once we require traitlets > 4.1 :-P

File: tools/secure_notebook.py [hunk 1]
```diff
@@ -19,13 +27,14 @@
...
 
 
-def create_self_signed_cert(cert_dir, keyfile, certfiile):
+def create_self_signed_cert(cert_dir, keyfile, certfile):
     """
-    If datacard.crt and datacard.key don't exist in cert_dir, create a new
-    self-signed cert and keypair and write them into that directory.
+    Create a self-signed `keyfile` and `certfile` in `cert_dir`
+
+    Abort if one of the
```
Comment (takluyver) [inline tools/secure_notebook.py]: of the...

File: tools/secure_notebook.py [hunk 2]
```diff
@@ -51,14 +60,17 @@ def create_self_signed_cert(cert_dir, keyfile, certfiile):
...
 
-if __name__ == '__main__':
-    print("This guide you into securing your notebook server")
-    print("first choose a password.")
-    pw = passwd()
-    print("We will store your password encrypted in the notebook configuration file: ")
-    print(pw)
+
+@contextmanager
+def load_config():
```
Comment (takluyver) [inline tools/secure_notebook.py]: This could be more clearly named to indicate that it's both loading and saving - maybe `persistent_config`?

File: tools/secure_notebook.py [hunk 3]
```diff
@@ -66,16 +78,33 @@ def create_self_signed_cert(cert_dir, keyfile, certfiile):
     except ConfigFileNotFound:
         config = Config()
 
-    config.NotebookApp.password = pw
+    yield config
 
     with io.open(os.path.join(jupyter_config_dir(), 'jupyter_notebook_config.json'), 'w') as f:
```
Comment (takluyver) [inline tools/secure_notebook.py]: If we're storing passwords, should we be trying to ensure that the config file is only user readable (i.e. chmod 600 on it)?

File: tools/secure_notebook.py [hunk 4]
```diff
@@ -66,16 +78,33 @@ def create_self_signed_cert(cert_dir, keyfile, certfiile):
...
+        config.NotebookApp.password = pw
+
+    print('... done\n')
+
+
+def set_certifs():
+    """
+    generate certificate to run notebook over ssl and set up the notebook config.
+    """
+    print("Let's generate self-signed certificates to secure your connexion.")
```
Comment (takluyver) [inline tools/secure_notebook.py]: Will it be practical to add an option here to generate a certificate with letsencrypt, once it's generally available, rather than a self-signed one?

File: tools/secure_notebook.py (hunk 3 above)
Comment (Carreau) [inline tools/secure_notebook.py]: That usually is handled at jupyter_ore level that config dir should be 0o700, the password is sha-hashed, so not **that** sensible.

File: tools/secure_notebook.py (hunk 4 above)
Comment (Carreau) [inline tools/secure_notebook.py]: Likely, in the end this is in tools, we could migrate it to the notebook itself later, but it does not replace a real audited way of having a SSL certificate and secure connexion.

File: tools/secure_notebook.py [hunk 5]
```diff
@@ -66,16 +78,33 @@ def create_self_signed_cert(cert_dir, keyfile, certfiile):
...
         f.write(six.u(json.dumps(config, indent=2)))
 
-    print('... done')
-    print()
 
-    print("Now let's generate self-signed certificates to secure your connexion.")
+def set_password():
+    """Ask user for password, store it in notebook json configuration file"""
+
+    print("first choose a password.")
```
Comment (fperez) [inline tools/secure_notebook.py]: Capitalize "First"

File: tools/secure_notebook.py [hunk 6]
```diff
@@ -66,16 +78,33 @@ def create_self_signed_cert(cert_dir, keyfile, certfiile):
...
+
+    with persist_config() as config:
+        config.NotebookApp.password = pw
+
+    print('... done\n')
+
+
+def set_certifs():
+    """
+    generate certificate to run notebook over ssl and set up the notebook config.
```
Comment (fperez) [inline tools/secure_notebook.py]: Capitalize "Generate"

File: tools/secure_notebook.py [hunk 7]
```diff
@@ -90,12 +119,14 @@ def create_self_signed_cert(cert_dir, keyfile, certfiile):
...
-    config.NotebookApp.keyfile = fullkey
-
-    with io.open(os.path.join(jupyter_config_dir(), 'jupyter_notebook_config.json'), 'w') as f:
-        f.write(six.u(json.dumps(config, indent=2)))
-
+    print('done.\n')
 
 
+if __name__ == '__main__':
+    print("This guide you into securing your notebook server.")
```
Comment (fperez) [inline tools/secure_notebook.py]: I'd reword the msg as `This will guide you through the steps towards securing your notebook server.` Though a review by one of our native speakers, @takluyver or @minrk would be even better...

File: tools/secure_notebook.py [hunk 8]
```diff
@@ -19,13 +27,14 @@
...
+def create_self_signed_cert(cert_dir, keyfile, certfile):
     """
-    If datacard.crt and datacard.key don't exist in cert_dir, create a new
-    self-signed cert and keypair and write them into that directory.
+    Create a self-signed `keyfile` and `certfile` in `cert_dir`
+
+    Abort if one of the keyfile of certfile exist.
     """
 
-    if not exists(join(cert_dir, certfiile)) \
```
Comment (fperez) [inline tools/secure_notebook.py]: I would write the logic in the opposite order: `if exists: raise, else: write()`. Less composite negatives make the code much easier to follow, I think, even if the exception goes first, which is a bit unusual.

But just my take on it, OK if you disagree.

File: tools/secure_notebook.py [hunk 9]
```diff
@@ -51,14 +60,17 @@ def create_self_signed_cert(cert_dir, keyfile, certfiile):
...
-    print("first choose a password.")
-    pw = passwd()
-    print("We will store your password encrypted in the notebook configuration file: ")
-    print(pw)
+
+@contextmanager
+def persist_config():
+    """Conext manager that can be use to modify a config object
+
+    on exit of the context manager, the config will be written back to disk.
```
Comment (fperez) [inline tools/secure_notebook.py]: Capitalize `On` at start of sentence.

File: tools/secure_notebook.py [hunk 10]
```diff
@@ -51,14 +60,17 @@ def create_self_signed_cert(cert_dir, keyfile, certfiile):
...
-if __name__ == '__main__':
-    print("This guide you into securing your notebook server")
-    print("first choose a password.")
-    pw = passwd()
-    print("We will store your password encrypted in the notebook configuration file: ")
-    print(pw)
+
+@contextmanager
+def persist_config():
+    """Conext manager that can be use to modify a config object
```
Comment (fperez) [inline tools/secure_notebook.py]: Typo: `Context`

File: tools/secure_notebook.py (hunk 3 above)
Comment (fperez) [inline tools/secure_notebook.py]: I think though that standard practice for these kinds of files is to still store them with tight permissions, even if they are in a 700 dir, in case they get temporarily/accidentally moved out.

File: tools/secure_notebook.py [hunk 11]
```diff
@@ -66,16 +78,33 @@ def create_self_signed_cert(cert_dir, keyfile, certfiile):
...
-    print()
 
-    print("Now let's generate self-signed certificates to secure your connexion.")
+def set_password():
+    """Ask user for password, store it in notebook json configuration file"""
+
+    print("first choose a password.")
+    pw = passwd()
+    print("We will store your password encrypted in the notebook configuration file: ")
+    print(pw)
```
Comment (fperez) [inline tools/secure_notebook.py]: Are we echoing passwords in plaintext back to the screen?

File: tools/secure_notebook.py (hunk 8 above)
Comment (Carreau) [inline tools/secure_notebook.py]: Done

File: tools/secure_notebook.py (hunk 10 above)
Comment (Carreau) [inline tools/secure_notebook.py]: Done

File: tools/secure_notebook.py (hunk 9 above)
Comment (Carreau) [inline tools/secure_notebook.py]: Done

File: tools/secure_notebook.py (hunk 3 above)
Comment (Carreau) [inline tools/secure_notebook.py]: persist config default to 0o600 now.

File: tools/secure_notebook.py (hunk 5 above)
Comment (Carreau) [inline tools/secure_notebook.py]: Done.

File: tools/secure_notebook.py (hunk 11 above)
Comment (Carreau) [inline tools/secure_notebook.py]: No hashed password, I changed the name of the variable.

File: tools/secure_notebook.py (hunk 6 above)
Comment (Carreau) [inline tools/secure_notebook.py]: Done

File: tools/secure_notebook.py (hunk 7 above)
Comment (Carreau) [inline tools/secure_notebook.py]: Done.

**Comment by fperez** [comment]
OK, I suggest we merge this now. It's a tool, so not in the critical path, and it's seen by now enough review/iteration that I think we can call it good enough for now.

Any objection?

**Comment by fperez** [comment]
OK, nobody objected, merging to move the queue forward.
